pgrx/src/fcall.rs

```diff
@@ -0,0 +1,143 @@
+            //         I think we'll need to fully port Postgres' "List" type to Rust...
+            let mut parts_list = PgList::new();
+            ident_parts
+                .unwrap()
+                .iter_deny_null()
+                .map(|part| pg_sys::makeString(part.as_pg_cstr()))
+                .for_each(|part| parts_list.push(part));
+
+            // ask Postgres to find the function.  This will look for the possibly-qualified named
+            // function following the normal SEARCH_PATH rules, ensuring its argument type Oids
+            // exactly match the ones from the user's input arguments.  It does not evaluate the
+            // return type, so we'll have to do that later
+            let func_oid = pg_sys::LookupFuncName(
+                parts_list.as_ptr(),
+                args.len().try_into().unwrap(),
+                arg_types.as_ptr(),
+                false,
+            );
+
+            // free the individual `pg_sys::String` parts we allocated above
+            parts_list.iter_ptr().for_each(|s| {
+                #[cfg(any(
+                    feature = "pg11",
+                    feature = "pg12",
+                    feature = "pg13",
+                    feature = "pg14"
+                ))]
+                pg_sys::pfree((*s).val.str_.cast());
+
+                #[cfg(any(feature = "pg15", feature = "pg16"))]
+                pg_sys::pfree((*s).sval.cast());
+            });
+
+            func_oid
+        };
+
+        //
+        // TODO:  lookup the function's pg_proc entry and do some validation around, at least,
+        //  - STRICT: If the function is STRICT, any `None` argument Datum should return None right now
+        //  - IN/OUT/TABLE arg types: I think we should only support IN in v1 of this
+        //  - Make sure the return type `T` is compatible with the resolved function's return type
+        //  - ??
+        //
+
+        //
+        // The following code is Postgres-version specific.  Right now, it's compatible with v12+
+        // v11 will need a different implementation.
+        //
+        // NB:  Which I don't want to do since it EOLs in 3 months
+        //
+
+        // initialize a stack-allocated `FmgrInfo` instance
+        let mut flinfo = pg_sys::FmgrInfo::default();
+        pg_sys::fmgr_info(func_oid, &mut flinfo);
+
+        // heap allocate a `FunctionCallInfoBaseData` properly sized so there's enough room
+        // for `args.len()` arguments
+        let fcinfo = pg_sys::palloc0(
+            std::mem::size_of::<pg_sys::FunctionCallInfoBaseData>()
+                + std::mem::size_of::<pg_sys::NullableDatum>() * args.len(),
+        ) as *mut pg_sys::FunctionCallInfoBaseData;
+
+        // initialize it
+        let fcinfo_ref = fcinfo.as_mut().unwrap();
+        fcinfo_ref.flinfo = &mut flinfo;
+        fcinfo_ref.fncollation = pg_sys::InvalidOid; // TODO:  We need to get a collation from somewhere?
```

Comment:
Ah, I have an excuse to follow up on this tomorrow:
- https://github.com/pgcentralfoundation/pgrx/issues/1257

Reply:
The answer to that issue's question is no. But also the answer to "what do we do here?" seems to be that, in most cases, the "direct function call" frameworks inside Postgres either explicitly pass a collation or assume InvalidOid is an acceptable input collation. There is a TODO to use `DEFAULT_COLLATION_OID` in a few such places, however.